File: typescript/src/dtn/IDtnBundleStore.ts
/**
 * DTN bundle store + in-memory default.
 * SPDX-License-Identifier: MIT
 */

import {
  BundleStatus,
  CustodyRecord,
  DtnBundle,
  isBundleExpired,
} from "../models/index.js";

export interface IDtnBundleStore {
  get(bundleId: string): Promise<DtnBundle | null>;
  getActive(): Promise<DtnBundle[]>;
  save(bundle: DtnBundle): Promise<void>;
  remove(bundleId: string): Promise<void>;
  getActiveCount(): Promise<number>;
  saveCustody(record: CustodyRecord): Promise<void>;
  getCustodyRecords(bundleId: string): Promise<CustodyRecord[]>;
  expireStale(): Promise<number>;
}

export class InMemoryDtnBundleStore implements IDtnBundleStore {
  private readonly bundles = new Map<string, DtnBundle>();
  private readonly custody = new Map<string, CustodyRecord>();

  async get(bundleId: string): Promise<DtnBundle | null> {
    return this.bundles.get(bundleId) ?? null;
  }

  async getActive(): Promise<DtnBundle[]> {
    const out: DtnBundle[] = [];
    for (const b of this.bundles.values()) {
      if (
        !isBundleExpired(b) &&
        (b.status === BundleStatus.Pending || b.status === BundleStatus.InCustody)
      ) {
        out.push(b);
      }
    }
    return out;
  }

  async save(bundle: DtnBundle): Promise<void> {
    this.bundles.set(bundle.id, bundle);
  }

  async remove(bundleId: string): Promise<void> {
    this.bundles.delete(bundleId);
  }

  async getActiveCount(): Promise<number> {
    let count = 0;
    for (const b of this.bundles.values()) {
      if (
        !isBundleExpired(b) &&
        (b.status === BundleStatus.Pending || b.status === BundleStatus.InCustody)
      ) {
        count++;
      }
    }
    return count;
  }

  async saveCustody(record: CustodyRecord): Promise<void> {
    this.custody.set(record.id, record);
  }

  async getCustodyRecords(bundleId: string): Promise<CustodyRecord[]> {
    return Array.from(this.custody.values()).filter((r) => r.bundleId === bundleId);
  }

  async expireStale(): Promise<number> {
    let expired = 0;
    for (const b of this.bundles.values()) {
      if (isBundleExpired(b) && b.status !== BundleStatus.Expired) {
        b.status = BundleStatus.Expired;
        expired++;
      }
    }
    return expired;
  }
}

File: src/AetherNet.Core/Dtn/IDtnBundleStore.cs
// SPDX-License-Identifier: MIT

using AetherNet.Models;

namespace AetherNet.Dtn;

/// <summary>
/// Persistent backing store for DTN bundles and their custody records.
/// Production hosts substitute SQLite- or file-backed implementations so bundles
/// survive restarts; the default <see cref="InMemoryDtnBundleStore"/> is process-local.
/// </summary>
public interface IDtnBundleStore
{
    /// <summary>Returns the bundle with the given id, or null if absent.</summary>
    Task<DtnBundle?> GetAsync(Guid bundleId, CancellationToken cancellationToken = default);

    /// <summary>Returns every bundle whose <see cref="DtnBundle.Status"/> is <see cref="BundleStatus.Pending"/> or <see cref="BundleStatus.InCustody"/>.</summary>
    Task<IReadOnlyList<DtnBundle>> GetActiveAsync(CancellationToken cancellationToken = default);

    /// <summary>Inserts or replaces a bundle by id.</summary>
    Task SaveAsync(DtnBundle bundle, CancellationToken cancellationToken = default);

    /// <summary>Removes the bundle with the given id, if present.</summary>
    Task RemoveAsync(Guid bundleId, CancellationToken cancellationToken = default);

    /// <summary>Returns the count of active (non-delivered, non-expired) bundles currently held.</summary>
    Task<int> GetActiveCountAsync(CancellationToken cancellationToken = default);

    /// <summary>Records a custody transfer.</summary>
    Task SaveCustodyAsync(CustodyRecord record, CancellationToken cancellationToken = default);

    /// <summary>Returns every custody record associated with the given bundle.</summary>
    Task<IReadOnlyList<CustodyRecord>> GetCustodyRecordsAsync(Guid bundleId, CancellationToken cancellationToken = default);

    /// <summary>Marks every expired bundle as <see cref="BundleStatus.Expired"/> and returns the count affected.</summary>
    Task<int> ExpireStaleAsync(CancellationToken cancellationToken = default);
}
